### PR TITLE
Fix locations reporting

### PR DESCRIPTION
**Pull request by SaphireLattice**
It was a bit of a mess, just a bit

### CODE REVIEW

File: DMCompiler/Compiler/Lexer.cs [hunk 1]
```diff
@@ -4,16 +4,20 @@
 
 internal class Lexer<TSourceType> {
     public Location CurrentLocation { get; protected set; }
-    public IEnumerable<TSourceType> Source { get; }
-    public bool AtEndOfSource { get; private set; }
+    public Location PreviousLocation { get; protected set; }
+    public string SourceName { get; protected set; }
+    public IEnumerable<TSourceType> Source { get; protected set; }
+    public bool AtEndOfSource { get; protected set; } = false;
```
[the remark this answers is not in the document]
Comment (SaphireLattice) [reply DMCompiler/Compiler/Lexer.cs:10]: Would have been nice if it reported the looser protection level too

File: DMCompiler/Compiler/DM/AST/DMAST.cs [hunk 2]
```diff
@@ -4,6 +4,21 @@ namespace DMCompiler.Compiler.DM.AST;
...
 public abstract class DMASTNode(Location location) {
     public readonly Location Location = location;
+
+    public override string ToString() {
+        return $"{ToString(null)}";
+    }
+
+    public string ToString(Location? loc, bool hideLoc = false) {
+        if (hideLoc || (loc is not null && Location.SourceFile == loc.Value.SourceFile && Location.Line == loc.Value.Line))
+            return $"{ToStringNoLocation()}";
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DM/AST/DMAST.cs:14]:
```suggestion
            return ToStringNoLocation();
```

File: DMCompiler/Compiler/DM/AST/DMAST.cs [hunk 3]
```diff
@@ -4,6 +4,21 @@ namespace DMCompiler.Compiler.DM.AST;
 
 public abstract class DMASTNode(Location location) {
     public readonly Location Location = location;
+
+    public override string ToString() {
+        return $"{ToString(null)}";
+    }
+
+    public string ToString(Location? loc, bool hideLoc = false) {
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DM/AST/DMAST.cs:12]: I don't see `hideLoc` used anywhere, and it seems redundant when you can give `null` for the first argument

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: DMCompiler/Compiler/DM/AST/DMAST.cs (hunk 3 above)
Comment (SaphireLattice) [reply DMCompiler/Compiler/DM/AST/DMAST.cs:12]: Ah forgot about null

And I've tried to uh, make the thing useful for debugging and stuff. Used it with the param when dealing with tokens and shit.

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**